hazelcast/src/main/java/com/hazelcast/crdt/pncounter/PNCounterService.java

```diff
@@ -0,0 +1,256 @@
+        final HashMap<String, PNCounterImpl> counters = new HashMap<String, PNCounterImpl>();
+        final Config config = nodeEngine.getConfig();
+
+        for (Entry<String, PNCounterImpl> counterEntry : this.counters.entrySet()) {
+            final String counterName = counterEntry.getKey();
+            final PNCounterImpl counter = counterEntry.getValue();
+            if (targetIndex >= config.findPNCounterConfig(counterName).getReplicaCount()) {
+                continue;
+            }
+            final VectorClock counterCurrentVectorClock = counter.getCurrentVectorClock();
+            final VectorClock counterPreviousVectorClock = previouslyReplicatedVectorClocks.get(counterName);
+
+            if (counterPreviousVectorClock == null || counterCurrentVectorClock.isAfter(counterPreviousVectorClock)) {
+                counters.put(counterName, counter);
+            }
+            currentVectorClocks.put(counterName, counterCurrentVectorClock);
+        }
+
+        return counters.isEmpty()
+                ? null
+                : new CRDTReplicationContainer(new PNCounterReplicationOperation(counters), currentVectorClocks);
+    }
+
+    @Override
+    public String getName() {
+        return SERVICE_NAME;
+    }
+
+    @Override
+    public void merge(String name, PNCounterImpl value) {
+        getCounter(name).merge(value);
+    }
+
+    @Override
+    public CRDTReplicationContainer prepareMigrationOperation(int maxConfiguredReplicaCount) {
+        final HashMap<String, VectorClock> currentVectorClocks = new HashMap<String, VectorClock>();
+        final HashMap<String, PNCounterImpl> counters = new HashMap<String, PNCounterImpl>();
+        final Config config = nodeEngine.getConfig();
+
+        for (Entry<String, PNCounterImpl> counterEntry : this.counters.entrySet()) {
+            final String counterName = counterEntry.getKey();
+            final PNCounterImpl counter = counterEntry.getValue();
+            if (config.findPNCounterConfig(counterName).getReplicaCount() >= maxConfiguredReplicaCount) {
+                continue;
+            }
+            counters.put(counterName, counter);
+            currentVectorClocks.put(counterName, counter.getCurrentVectorClock());
+        }
+
+        return counters.isEmpty()
+                ? null
+                : new CRDTReplicationContainer(new PNCounterReplicationOperation(counters), currentVectorClocks);
+    }
+
+    @Override
+    public boolean clearCRDTState(Map<String, VectorClock> vectorClocks) {
+        boolean allCleared = true;
+        for (Entry<String, VectorClock> vectorClockEntry : vectorClocks.entrySet()) {
+            final String counterName = vectorClockEntry.getKey();
+            final VectorClock vectorClock = vectorClockEntry.getValue();
+            final PNCounterImpl counter = counters.get(counterName);
+            if (counter == null) {
```

Comment:
Seems like this is possible only when someone destroyed the counter while we are trying to migrate it. In this case it may survive on another member because it may be migrated after the destroy operation is already performed on that member. If I understand things correctly, this may lead to a formation of flaky counters which are present only on some nodes.

Reply:
That's one way. One other way would be that there are two concurrent migration operations - e.g. there was a merge followed by an membership event on the same replica. These will start two migration tasks but only one can remove a counter from the map.

As for the concurrent destroy, true, the counter may survive on one node. In that case, it will be again replicated to other members, since from the replication service we on the next replication round see that other members haven't observed the counter. So in a sense, the counter survived the destroy.

Reply:
Thank you for the clarification. This destroy issue doesn't look that critical unless one relies heavily on the existence of counters or "remaps" their names so they change the semantics of counters. Seems like those are not frequent use cases.